### PR TITLE
Refurbished logging module.

### PR DESCRIPTION
**Pull request by YooSunYoung**

### CODE REVIEW

File: tests/logging/contexts.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    within the sub context using ``log_factory``.
+
+    It will help a test not to interfere other tests.
+    """
+
+    original_logger_names = list(logging.Logger.manager.loggerDict.keys())
+    try:
+        with log_factory.local_factory(*local_log_providers) as local_log_factory:
+            yield local_log_factory
+    finally:
```
Comment (YooSunYoung) [inline tests/logging/contexts.py:28]: It cleans up any loggers that were added to the `logging.Logger.manager` within this context.
It was easier than resetting handlers of each loggers. It also keeps original loggers that are unrelated to the tests as they were.

File: src/beamlime/logging/resources.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    try:
+        yield _lock
+    finally:
+        _lock.release() if _lock else ...
+
+
+def cleanup_file_handlers(logger: logging.Logger):
+    """Find file handlers that are connectected to
+    non-existing files and remove them from the logger."""
+    from os.path import exists
```
Comment (YooSunYoung) [inline src/beamlime/logging/resources.py:110]: It probably should raise an error instead of just removing the handlers...

File: src/beamlime/logging/resources.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+def create_time_stamp() -> TimeStamp:
+    """
+    Creates a timestamp with the format, ``{cur_timestamp}--{h_time}``.
+    ``htime`` is as an extra information only for the human users,
+    and the ``cur_timestamp`` should be used by the applications or modules.
+    """
+    from datetime import datetime
+
+    cur_timestamp = round(time.time())
+    h_time = datetime.fromtimestamp(cur_timestamp).strftime("%Y-%m-%d-%H-%M-%S")
```
Comment (jl-wynen) [inline src/beamlime/logging/resources.py:27]: Can you change this to use the ISO format?

```suggestion
    h_time = datetime.fromtimestamp(cur_timestamp).isoformat()
```

And ideally even with timezone information. I'm weary of times without timezone information as they can easily be misinterpreted.
You could use the local timezone if you are willing to use a package like dateutil. But I think it is enough to show everything in UTC using:

```suggestion
    from datetime import datetime, timezone

    now = datetime.now(tz=timezone.utc)
    cur_timestamp = round(now.timestamp())
    h_time = now.isoformat()
```

But in general, why do we need the timestamp? If we use the full ISO format, it can be parsed back into a `datetime`, esp. if we force it to be in UTC.

File: src/beamlime/logging/resources.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    return TimeStamp(f"{cur_timestamp}--{h_time}")
+
+
+LogFilePrefix = NewType("LogFilePrefix", str)
+DefaultPrefix = LogFilePrefix("beamlime")
+
+LogFileExtension = NewType("LogFileExtension", str)
+DefaultFileExtension = LogFileExtension("log")
+
+LogDirectoryPath = NewType("LogDirectoryPath", str)
```
Comment (jl-wynen) [inline src/beamlime/logging/resources.py:37]: Why not use `Path` instead of `str` here? Then you can even use methods on it instead of the functions in `os`.

File: src/beamlime/logging/resources.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+
+@log_providers.provider
+def create_log_file_name(
+    time_stamp: TimeStamp,
+    prefix: LogFilePrefix = DefaultPrefix,
+    suffix: LogFileExtension = DefaultFileExtension,
+) -> LogFileName:
+    """Combine prefix, timestamp and suffix into a new file name."""
+    return LogFileName(f"{prefix}--{time_stamp}.{suffix}")
```
Comment (jl-wynen) [inline src/beamlime/logging/resources.py:70]: Did you check if this works on Windows? I don't know if it supports numbers and dashes in filenames like this.

File: src/beamlime/logging/resources.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+
+FileHandlerBasePath = NewType("FileHandlerBasePath", Path)
+DefaultDirectoryStatus = DirectoryCreated(True)
+
+
+@log_providers.provider
+def create_log_file_path(
+    directory_ready: DirectoryCreated = DefaultDirectoryStatus,
```
Comment (jl-wynen) [inline src/beamlime/logging/resources.py:79]: Why is there a default argument?

File: src/beamlime/logging/resources.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+@contextmanager
+def hold_logging() -> Iterator[RLock]:
+    from . import _lock
+
+    _lock.acquire() if _lock else ...
+    try:
+        yield _lock
+    finally:
+        _lock.release() if _lock else ...
```
Comment (jl-wynen) [inline src/beamlime/logging/resources.py:104]: What are the `if _lock` checks for? And why does this function yield the lock?

File: src/beamlime/logging/resources.py (hunk 4 above)
Comment (YooSunYoung) [reply src/beamlime/logging/resources.py:37]: That is very true... I thought it'll be easier to keep it as `str` in case it'll be user input later but I can just wrap the user input with `Path` so it doesn't matter. I'll update it...!

File: src/beamlime/logging/resources.py (hunk 7 above)
Comment (YooSunYoung) [reply src/beamlime/logging/resources.py:104]: I just tried to keep it consistent with what `logging` library does. But it's no longer used in the `beamlime` anymore so I'll just remove it...! Thank you!

File: src/beamlime/logging/resources.py (hunk 6 above)
Comment (YooSunYoung) [reply src/beamlime/logging/resources.py]: Actually no... I removed the default argument. Thank you!

File: src/beamlime/logging/resources.py (hunk 5 above)
Comment (YooSunYoung) [reply src/beamlime/logging/resources.py]: Actually no... for now the file path doesn't have `--` anymore but I should probably add `Windows` into CI tests.
I opened an issue #72

File: src/beamlime/logging/__init__.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+    BeamlimeLogger,
+    FileHandlerConfigured,
+    LogLevels,
+    get_logger,
+    get_scipp_logger,
+    initialize_file_handler,
+)
+from .resources import LogDirectoryPath, LogFileExtension, LogFileName, LogFilePrefix
+
+_lock = threading.RLock()
```
[the remark this answers is not in the document]
Comment (YooSunYoung) [reply src/beamlime/logging/__init__.py]: Right... I forgot that it doesn't check  `__init__.py` for unused variables...

**Review by jl-wynen — APPROVED** [review]
